blockchain/chainio.go

```diff
@@ -1116,48 +1117,142 @@ func dbFetchUtxoStats(dbTx database.Tx) (*UtxoStats, error) {
+	}
+	key := outpointKey(outpoint)
+	err = utxoBucket.Put(*key, serialized)
+	// NOTE: The key is intentionally not recycled here since the database
+	// interface contract prohibits modifications.  It will be garbage collected
+	// normally when the database is done with it.
+	if err != nil {
+		return err
 	}
 
 	return nil
 }
 
+// -----------------------------------------------------------------------------
+// The utxo set state contains information regarding the current state of the
+// utxo set.  In particular, it tracks the block height and block hash of the
+// last completed flush.
+//
+// The utxo set state is tracked in the database since at any given time, the
+// utxo cache may not be consistent with the utxo set in the database.  This is
+// due to the fact that the utxo cache only flushes changes to the database
+// periodically.  Therefore, during initialization, the utxo set state is used
+// to identify the last flushed state of the utxo set and it can be caught up
+// to the current best state of the main chain.
+//
+// Note: The utxo set state MUST always be updated in the same database
+// transaction that the utxo set is updated in to guarantee that they stay in
+// sync in the database.
+//
+// The serialized format is:
+//
+//   <block height><block hash>
+//
+//   Field          Type             Size
+//   block height   VLQ              variable
+//   block hash     chainhash.Hash   chainhash.HashSize
+//
+// -----------------------------------------------------------------------------
+
+// utxoSetState represents the current state of the utxo set.  In particular,
+// it tracks the block height and block hash of the last completed flush.
+type utxoSetState struct {
+	lastFlushHeight uint32
+	lastFlushHash   chainhash.Hash
+}
+
+// serializeUtxoSetState serializes the provided utxo set state.  The format is
+// described in detail above.
+func serializeUtxoSetState(state *utxoSetState) []byte {
+	// Calculate the size needed to serialize the utxo set state.
+	size := serializeSizeVLQ(uint64(state.lastFlushHeight)) + chainhash.HashSize
+
+	// Serialize the utxo set state and return it.
+	serialized := make([]byte, size)
+	offset := putVLQ(serialized, uint64(state.lastFlushHeight))
+	copy(serialized[offset:], state.lastFlushHash[:])
+	return serialized
+}
+
+// deserializeUtxoSetState deserializes the passed serialized byte slice into
+// the utxo set state.  The format is described in detail above.
+func deserializeUtxoSetState(serialized []byte) (*utxoSetState, error) {
+	// Deserialize the block height.
+	blockHeight, bytesRead := deserializeVLQ(serialized)
+	offset := bytesRead
+	if offset >= len(serialized) {
+		return nil, errDeserialize("unexpected end of data after height")
+	}
+
+	// Deserialize the hash.
+	if len(serialized[offset:]) != chainhash.HashSize {
+		return nil, errDeserialize("unexpected length for serialized hash")
+	}
+	var hash chainhash.Hash
+	copy(hash[:], serialized[offset:chainhash.HashSize])
```

Comment:
This is incorrect.  The offset will be non-zero here and so it will end up chopping off part of the hash leaving a bunch of trailing zeroes (that then end up showing as leading zeroes when it gets treated as a LE uint256 ).

```suggestion
	copy(hash[:], serialized[offset:offset+chainhash.HashSize])
```

Reply:
Good catch, this is resolved now.  Missed this in testing with mainnet (and using a mainnet hash in the tests) due to the leading zeros.  Added a test case that would have failed previously as well.

Reply:
Confirmed this resolves the issue.